src/main/java/graphql/parser/Parser.java

```diff
@@ -227,6 +237,31 @@ public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int
         return node;
     }
 
+    private void setupParserListener(MultiSourceReader multiSourceReader, GraphqlParser parser, GraphqlAntlrToLanguage toLanguage) {
+        int maxTokens = toLanguage.getParserOptions().getMaxTokens();
+        // prevent a billion laugh attacks by restricting how many tokens we allow
+        ParseTreeListener listener = new GraphqlBaseListener() {
+            int count = 0;
+
+            @Override
+            public void visitTerminal(TerminalNode node) {
+                count++;
+                if (count > maxTokens) {
+                    String msg = String.format("More than %d parse tokens have been presented.  To prevent Denial Of Service attacks, parsing has been cancelled.", maxTokens);
```

Comment:
```suggestion
                    String msg = String.format("More than %d parse tokens have been presented. To prevent Denial Of Service attacks, parsing has been cancelled.", maxTokens);
```